common/client/transaction_sender.go

```diff
@@ -0,0 +1,261 @@
+	SendTransaction(ctx context.Context, tx TX) error
+}
+
+func NewTransactionSender[TX any, CHAIN_ID types.ID, RPC SendTxRPCClient[TX]](
+	lggr logger.Logger,
+	chainID CHAIN_ID,
+	chainFamily string,
+	multiNode *MultiNode[CHAIN_ID, RPC],
+	txErrorClassifier TxErrorClassifier[TX],
+	sendTxSoftTimeout time.Duration,
+) TransactionSender[TX] {
+	if sendTxSoftTimeout == 0 {
+		sendTxSoftTimeout = QueryTimeout / 2
+	}
+	return &transactionSender[TX, CHAIN_ID, RPC]{
+		chainID:           chainID,
+		chainFamily:       chainFamily,
+		lggr:              logger.Sugared(lggr).Named("TransactionSender").With("chainID", chainID.String()),
+		multiNode:         multiNode,
+		txErrorClassifier: txErrorClassifier,
+		sendTxSoftTimeout: sendTxSoftTimeout,
+	}
+}
+
+type transactionSender[TX any, CHAIN_ID types.ID, RPC SendTxRPCClient[TX]] struct {
+	services.StateMachine
+	chainID           CHAIN_ID
+	chainFamily       string
+	lggr              logger.SugaredLogger
+	multiNode         *MultiNode[CHAIN_ID, RPC]
+	txErrorClassifier TxErrorClassifier[TX]
+	sendTxSoftTimeout time.Duration // defines max waiting time from first response til responses evaluation
+
+	// TODO: add start/ stop methods. Start doesn't need to do much.
+	// TODO: Stop should stop sending transactions, and close chStop to stop collecting results, reporting/ etc.
+	chStop services.StopChan
+}
+
+// SendTransaction - broadcasts transaction to all the send-only and primary nodes in MultiNode.
+// A returned nil or error does not guarantee that the transaction will or won't be included. Additional checks must be
+// performed to determine the final state.
+//
+// Send-only nodes' results are ignored as they tend to return false-positive responses. Broadcast to them is necessary
+// to speed up the propagation of TX in the network.
+//
+// Handling of primary nodes' results consists of collection and aggregation.
+// In the collection step, we gather as many results as possible while minimizing waiting time. This operation succeeds
+// on one of the following conditions:
+// * Received at least one success
+// * Received at least one result and `sendTxSoftTimeout` expired
+// * Received results from the sufficient number of nodes defined by sendTxQuorum.
+// The aggregation is based on the following conditions:
+// * If there is at least one success - returns success
+// * If there is at least one terminal error - returns terminal error
+// * If there is both success and terminal error - returns success and reports invariant violation
+// * Otherwise, returns any (effectively random) of the errors.
+func (txSender *transactionSender[TX, CHAIN_ID, RPC]) SendTransaction(ctx context.Context, tx TX) (SendTxReturnCode, error) {
+	txResults := make(chan sendTxResult, len(txSender.multiNode.primaryNodes))
+	txResultsToReport := make(chan sendTxResult, len(txSender.multiNode.primaryNodes))
+	primaryWg := sync.WaitGroup{}
+
+	var err error
+	ok := txSender.multiNode.IfNotStopped(func() {
```

Comment:
We should should not rely on multiNode's `StateMachine` as it's a separate component. 
Use `chStop` to track if we need to continue processing request and  `Close` with `WaitGroup` to signal when we are done to components that depend on us.